### PR TITLE
Args

### PR DESCRIPTION
**Pull request by albertjan**
some more pegen funcs

### CODE REVIEW

**Comment by albertjan** [comment]
# 426!

**Comment by albertjan** [comment]
# 430! 🎉

**Comment by albertjan** [comment]
# 476! 🚀

**Comment by albertjan** [comment]
# 481 🚂

**Comment by albertjan** [comment]
these are failing:
```
	t086.py
	t161.py
	t162.py
	t167.py
	t168.py
	t169.py
	t170.py
	t171.py
	t174.py
	t200.py
	t201.py
	t210.py
	t214.py
	t215.py
	t217.py
	t235.py
	t236.py
	t237.py
	t246.py
	t265.py
	t272.py
	t277.py
	t278.py
	t279.py
	t280.py
	t282.py
	t283.py
	t284.py
	t285.py
	t290.py
	t322.py
	t338.py
	t345.py
	t374.py
	t401.py
	t402.py
	t403.py
	t404.py
	t409.py
	t411.py
	t422.py
	t429.py
	t430.py
	t433.py
	t438.py
	t439.py
	t440.py
	t443.py
	t446.py
	t452.py
	t462.py
	t468.py
	t477.py
	t478.py
	t479.py
	t488.py
	t491.py
	t492.py
	t498.py
	t505.py
	t506.py
	t511.py
	t512.py
	t513.py
	t517.py
	t524.py
	t528.py
	t532.py
	t535.py
	t539.py
	t542.py
	t543.py
	t544.py
	t545.py
	t546.py
	t551.py
	t553.py
	t554.py
	t555.py
	```

**Comment by albertjan** [comment]
# 491

**Comment by albertjan** [comment]
# 500 💯💯💯💯💯

**Comment by s-cork** [comment]
wow this is amazing! sorry had a busy evening will have a look in a bit - but wow!

**Comment by albertjan** [comment]
I had lots of meeting I could ignore 😄

**Comment by albertjan** [comment]
failing now:

```
	t086.py
	t161.py
	t162.py
	t167.py
	t168.py
	t169.py
	t170.py
	t171.py
	t174.py
	t200.py
	t201.py
	t210.py
	t214.py
	t215.py
	t235.py
	t236.py
	t237.py
	t246.py
	t265.py
	t272.py
	t277.py
	t278.py
	t283.py
	t284.py
	t290.py
	t322.py
	t338.py
	t345.py
	t374.py
	t401.py
	t402.py
	t403.py
	t404.py
	t409.py
	t411.py
	t422.py
	t429.py
	t430.py
	t433.py
	t438.py
	t439.py
	t440.py
	t443.py
	t446.py
	t452.py
	t462.py
	t468.py
	t477.py
	t488.py
	t492.py
	t498.py
	t505.py
	t511.py
	t513.py
	t517.py
	t524.py
	t528.py
	t535.py
	t539.py
	t542.py
	```

File: src/parser/pegen.ts [hunk 1]
```diff
@@ -1502,6 +1521,23 @@ export function seq_flatten(p: Parser, seqs: AST[][]): AST[] {
 //     return _Py_Name(uni, Load, EXTRA_EXPR(first_name, second_name));
 // }
 
+class UnreachableException extends Error {}
```
Comment (s-cork) [inline src/parser/pegen.ts:1524]: Errors is another layer we need to think about at some point.

File: src/parser/pegen.ts [hunk 2]
```diff
@@ -1502,6 +1521,23 @@ export function seq_flatten(p: Parser, seqs: AST[][]): AST[] {
...
+
+export function seq_count_dots(seq: TokenInfo[]) {
```
Comment (s-cork) [inline src/parser/pegen.ts:1526]: I don' t know why but this seems quite a bizarre function - you know you've spent to long on a parser when you're counting dots.

File: src/parser/pegen.ts [hunk 3]
```diff
@@ -2118,7 +2239,7 @@ kwarg_or_starred[KeywordOrStarred*]:
     | a=starred_expression { _PyPegen_keyword_or_starred(p, a, 0) } <<-- here
     | invalid_kwarg
 */
-export function keyword_or_starred(p: Parser, element: any, is_keyword: boolean) {
+export function keyword_or_starred(p: Parser, element: keyword, is_keyword: boolean) {
```
Comment (s-cork) [inline src/parser/pegen.ts:2242]: Maybe this is something we can inline?

File: src/parser/pegen.ts [hunk 4]
```diff
@@ -2149,8 +2270,12 @@ export function keyword_or_starred(p: Parser, element: any, is_keyword: boolean)
 //     return n;
 // }
 
-export function seq_extract_starred_exprs(p: Parser, kwargs: KeywordOrStarred[]) {
-    return kwargs.filter((k) => k.is_keyword);
+function isKeyword(kw: expr | keyword): kw is keyword {
+    return kw instanceof keyword;
+}
```
Comment (s-cork) [inline src/parser/pegen.ts:2275]: our first typeguard

**Review by s-cork — APPROVED** [review]

**Comment by s-cork** [comment]
```
Ran 281 tests in 28.982s

FAILED (failures=77, errors=12, skipped=4)
```
for the `vr test pypeg`

File: src/parser/generated_parser.ts [hunk 5]
```diff
@@ -2935,7 +2940,7 @@ export class GeneratedParser extends Parser {
         const mark = this.mark();
         if ((a = this.name()) && (literal = this.expect("=")) && (b = this.expression())) {
             const EXTRA = this.extra(mark);
-            return pegen.keyword_or_starred(this, CHECK(new astnodes.keyword(a.id, b, ...EXTRA)), 1);
+            return pegen.fkeyword_or_starred(this, CHECK(new astnodes.keyword(a.id, b, ...EXTRA)), 1);
```
Comment (s-cork) [inline src/parser/generated_parser.ts]: yeah i think i've accidentally changed this a few times thinking i must have put the f in by mistake

Comment (albertjan) [reply src/parser/generated_parser.ts]: HA!

File: src/parser/pegen.ts (hunk 2 above)
Comment (albertjan) [reply src/parser/pegen.ts:1526]: lol yeah

File: src/parser/pegen.ts (hunk 1 above)
Comment (albertjan) [reply src/parser/pegen.ts:1524]: Yeah I needed a error

**Comment by albertjan** [comment]
should we look into test coverage?

File: src/parser/pegen.ts (hunk 3 above)
Comment (albertjan) [reply src/parser/pegen.ts:2242]: yeah probably we can do away with this whole concept since we have reflection 😄

File: src/parser/pegen.ts (hunk 4 above)
Comment (albertjan) [reply src/parser/pegen.ts:2275]: `kw instanceof keyword` is probably a typeguard in itself

File: src/parser/pegen.ts [hunk 6]
```diff
@@ -1502,6 +1521,23 @@ export function seq_flatten(p: Parser, seqs: AST[][]): AST[] {
...
+            switch (current_expr.type) {
+                case ELLIPSIS:
+                    return 3;
+                case DOT:
+                    return 1;
+                default:
+                    throw new UnreachableException();
+            }
+        })
+        .reduce((a, b) => a + b, 0);
```
Comment (albertjan) [inline src/parser/pegen.ts:1538]: I was again disappointed in javascript for not having a `sum` function

**Comment by s-cork** [comment]
I think that the py peg tests will do a good job of test coverage - so I think we might want to start using that as the barometer - i just added a pegen function for classdef decorators i'll push here since i'm on this branch locally

**Comment by albertjan** [comment]
I'll see if I can add that to the comment on the PR because you're probably right.

File: src/parser/pegen.ts (hunk 6 above)
Comment (s-cork) [reply src/parser/pegen.ts:1539]: would it be the same to do 

```
function _getNumDots(e: TokenInfo) {
    switch (e.type) {
        case ELLIPSIS:
            return 3;
        case DOT:
            return 1;
        default:
            throw new UnreachableException();
    }
}

export function seq_count_dots(seq: TokenInfo[]) {
    return seq.reduce((a, b) => a + _getNumDots(b), 0)
}
```

Comment (albertjan) [reply src/parser/pegen.ts:1539]: yeah it would, but it is better? 😄

**Comment by s-cork** [comment]
fyi - there's a slight bug in that test file

when I run it the first time the first test fails
the second time it runs fine
some sort of race condition 🤷‍♂️

File: src/parser/pegen.ts (hunk 6 above)
Comment (s-cork) [reply src/parser/pegen.ts]: it means we don't iterate over the array twice. once for map and once for reduce

Comment (s-cork) [reply src/parser/pegen.ts]: 088b388 - feel free to revert

**Comment by s-cork** [comment]
And with the python tests it might be worth turning each test into a t style test. Maybe a p001.py or something. It's really nice to be able to write vr parse 2 and vr test 2. I haven't found vr test pypeg particularly easy to dev with.

**Comment by albertjan** [comment]
Alright sounds like it would be worth our time to invest in that a bit. Do you want to have a chat tomorrow my morning to talk about how to progress from here?

File: src/parser/pegen.ts (hunk 6 above)
Comment (albertjan) [reply src/parser/pegen.ts]: Ah yeah of course I always forget that these are not iterators

**Pull request merged**